go/libraries/doltcore/rowconv/super_schema.go

```diff
@@ -0,0 +1,267 @@
+	err := allCols.Iter(func(tag uint64, col schema.Column) (stop bool, err error) {
+		inNameToOutName[col.Name] = fmt.Sprintf("%d_%s", col.Tag, col.Kind.String())
+		return false, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	fm, err := NewFieldMappingFromNameMap(sch, super, inNameToOutName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return NewRowConverter(fm)
+}
+
+// TagKindPair is a simple tuple that holds a tag and a NomsKind of a column
+type TagKindPair struct {
+	// Tag is the tag of a column
+	Tag uint64
+
+	// Kind is the NomsKind of a colum
+	Kind types.NomsKind
+}
+
+// NameKindPair is a simple tuple that holds the name of a column and it's NomsKind
+type NameKindPair struct {
+	// Name is the name of the column
+	Name string
+
+	// Kind is the NomsKind of the column
+	Kind types.NomsKind
+}
+
+// SuperSchemaGen is a utility class used to generate the superset of several schemas.
+type SuperSchemaGen struct {
+	tagKindToDestTag map[TagKindPair]uint64
+	usedTags         map[uint64]struct{}
+}
+
+// NewSuperSchemaGen creates a new SuperSchemaGen
+func NewSuperSchemaGen() *SuperSchemaGen {
+	return &SuperSchemaGen{make(map[TagKindPair]uint64), make(map[uint64]struct{})}
+}
+
+// AddSchema will add a schema which will be incorporated into the superset of schemas
+func (ss *SuperSchemaGen) AddSchema(sch schema.Schema) error {
+	err := sch.GetAllCols().Iter(func(tag uint64, col schema.Column) (stop bool, err error) {
+		tagKind := TagKindPair{Tag: tag, Kind: col.Kind}
+		_, exists := ss.tagKindToDestTag[tagKind]
+
+		if !exists {
+			destTag := tag
+
+			for {
+				_, collides := ss.usedTags[destTag]
+				if !collides {
+					ss.tagKindToDestTag[tagKind] = destTag
+					ss.usedTags[destTag] = struct{}{}
+					return false, nil
+				}
+
+				if destTag == tag {
+					destTag = schema.ReservedTagMin
+				} else {
+					destTag++
+				}
+			}
+		}
+
+		return false, nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GenerateSuperSchema takes all the accumulated schemas and generates a schema which is the superset of all of
+// those schemas.
+func (ss *SuperSchemaGen) GenerateSuperSchema(additionalCols ...NameKindPair) (schema.Schema, error) {
+	colColl, _ := schema.NewColCollection()
+	for tagKind, destTag := range ss.tagKindToDestTag {
+		colName := fmt.Sprintf("%d_%s", tagKind.Tag, tagKind.Kind.String())
+		col := schema.NewColumn(colName, destTag, tagKind.Kind, false)
+
+		var err error
+		colColl, err = colColl.Append(col)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(additionalCols) > 0 {
+		nextReserved := schema.ReservedTagMin
+		for {
+			if _, ok := ss.usedTags[nextReserved]; !ok {
+				break
+			}
+			nextReserved++
+		}
+
+		for _, nameKindPair := range additionalCols {
+			var err error
+			colColl, err = colColl.Append(schema.NewColumn(nameKindPair.Name, nextReserved, nameKindPair.Kind, false))
+
+			if err != nil {
+				return nil, err
+			}
+
+			nextReserved++
```

Comment:
Technically you should have the more complicated tag logic above in here too, unless you can guarantee that the reserved tags will always be assigned sequentially

Reply:
fixed